crates/server/src/services/core.rs

```diff
@@ -1,297 +1,395 @@
+        self.inner.store.as_ref()
     }
 
-    /// Get the map data associated with the core service.
-    pub fn map_data(&self) -> &Arc<RwLock<MapData>> {
-        &self.map
+    /// Submits a package record to be processed.
+    pub async fn submit_package_record(&self, log_id: LogId, record_id: RecordId) {
+        self.submit_entry_tx
+            .send(LogLeaf { log_id, record_id })
+            .await
+            .unwrap()
     }
+}
+
+struct Inner<Digest: SupportedDigest> {
+    // Operator signing key
+    operator_key: PrivateKey,
+
+    // DataStore persists transparency state.
+    store: Box<dyn DataStore>,
+
+    // In-memory transparency state.
+    state: RwLock<State<Digest>>,
+}
+
+impl<Digest: SupportedDigest> Inner<Digest> {
+    // Load state from DataStore or initialize empty state, returning any
+    // entries that are not yet part of a checkpoint.
+    async fn initialize(&mut self) -> Result<Vec<LogLeaf>, CoreServiceError> {
+        let initial = self.store.get_initial_leaves().await?.peekable();
+        pin_mut!(initial);
+
+        // If there are no stored log entries, initialize a new state
+        if initial.as_mut().peek().await.is_none() {
+            self.initialize_new().await?;
+            return Ok(vec![]);
+        }
+
+        // Reconstruct internal state from previously-stored log entires
+        let state = self.state.get_mut();
+        let mut pending_entries = vec![];
+        let mut last_checkpoint = Hash::<Digest>::of(()).into();
 
-    async fn initialize(
-        signing_key: &PrivateKey,
-        store: &dyn DataStore,
-    ) -> Result<InitializationData, DataStoreError> {
-        tracing::debug!("initializing core service");
-        let mut data = InitializationData::default();
-        let mut last_checkpoint = None;
-        let mut initial = store.get_initial_leaves().await?;
         while let Some(res) = initial.next().await {
             let InitialLeaf { leaf, checkpoint } = res?;
-            data.log.push(&leaf);
-            data.map = data.map.insert(
-                leaf.log_id.clone(),
-                MapLeaf {
-                    record_id: leaf.record_id.clone(),
-                },
-            );
 
-            match checkpoint {
-                Some(checkpoint) => {
-                    if last_checkpoint.as_ref() != Some(&checkpoint) {
-                        data.map_data.insert(data.map.clone());
-                        last_checkpoint = Some(checkpoint);
-                    }
+            // Push the entry itself
+            state.push_entry(leaf.clone());
+
+            // Update per-checkpoint state and track any entries that aren't
+            // yet part of a checkpoint
+            if let Some(checkpoint) = checkpoint {
+                if last_checkpoint != checkpoint {
+                    state.checkpoint(); // for the side-effect of updating map_index
+                    last_checkpoint = checkpoint;
                 }
```

Comment:
I think this contains a bug copied over from my initial implementation of server restart that I haven't gotten around to fixing.

Let's say the initial stream looks like (in record insert order):

| Leaf | Checkpoint|
|------|------|
| A  |  C1 |
| B | C2 |
| C | C3 |
| D | C3 |
| E | C3 |
| F | C4 | 

I think this code would correctly create a checkpoint `C1` containing `A` (`C1 != <empty-hash>`) and a checkpoint `C2` containing `B` (`C2 != C1`), but would not correctly checkpoint `C3` because we're checkpointing at the _transition_ from leaf `B` to `C` (`C3 != C2`) in the stream, which results in only including leaf `C` in the checkpoint.

The next checkpoint would then incorrectly include leafs `D`, `E`, and `F` (`C4 != C3`).

Does that make sense or am I completely off-base?

Reply:
Ah yeah I was a little confused about this but wanted to stick with existing behavior. I'm currently working on a followup refactor that avoids this problem entirely, but I can retrofit this PR if you'd prefer.

Reply:
I'll approve and leave it up to you whether you want to include it now or wait to the refactor.